VitruvioHost/Plugins/Vitruvio/Source/VitruvioEditor/Private/VitruvioCooker.cpp

```diff
@@ -343,75 +354,135 @@ void CookVitruvioActors(TArray<AActor*> Actors)
+				UStaticMeshComponent* CookedMeshComponent = AttachMeshComponent<UStaticMeshComponent>(CookedActor, PersistedMesh, GeneratedModelStaticMeshComponent->GetFName(), GeneratedModelStaticMeshComponent->GetComponentTransform());
+
+				CookOverrideMaterials(GeneratedModelStaticMeshComponent, CookedMeshComponent);
+				
+				// Persist instanced Components
+				TArray<TObjectPtr<USceneComponent>> AttachedComponents = GeneratedModelStaticMeshComponent->GetAttachChildren();
+				for (UActorComponent* ActorComponent : AttachedComponents)
+				{
+					UGeneratedModelHISMComponent* GeneratedModelHismComponent = Cast<UGeneratedModelHISMComponent>(ActorComponent);
 
-						InstancedStaticMeshComponent->SetMaterial(MaterialIndex, NewMaterial);
+					if (!GeneratedModelHismComponent)
+					{
+						continue;
 					}
 
-					for (int32 InstanceIndex = 0; InstanceIndex < GeneratedModelHismComponent->GetInstanceCount(); ++InstanceIndex)
+					if (UStaticMesh* InstanceMesh = GeneratedModelHismComponent->GetStaticMesh())
 					{
-						FTransform Transform;
-						GeneratedModelHismComponent->GetInstanceTransform(InstanceIndex, Transform);
-						InstancedStaticMeshComponent->AddInstance(Transform);
+						if (InstanceMesh->GetOutermost() == GetTransientPackage())
+						{
+							UStaticMesh* PersistedInstanceMesh = SaveStaticMesh(InstanceMesh, CookPath, MeshCache, MaterialCache, TextureCache);
+
+							FName Name = MakeUniqueObjectName(CookedActor, UHierarchicalInstancedStaticMeshComponent::StaticClass(), *PersistedInstanceMesh->GetName());
+							auto* InstancedStaticMeshComponent = AttachMeshComponent<UHierarchicalInstancedStaticMeshComponent>(CookedActor, CookedMeshComponent, PersistedInstanceMesh, Name, GeneratedModelHismComponent->GetComponentTransform());
+
+							CookOverrideMaterials(GeneratedModelHismComponent, InstancedStaticMeshComponent);
+
+							for (int32 InstanceIndex = 0; InstanceIndex < GeneratedModelHismComponent->GetInstanceCount(); ++InstanceIndex)
+							{
+								FTransform Transform;
+								GeneratedModelHismComponent->GetInstanceTransform(InstanceIndex, Transform);
+								InstancedStaticMeshComponent->AddInstance(Transform);
+							}
+						}
+						else
+						{
+							FName Name = MakeUniqueObjectName(CookedActor, UHierarchicalInstancedStaticMeshComponent::StaticClass(), *InstanceMesh->GetName());
+							auto* InstancedStaticMeshComponent = AttachMeshComponent<UHierarchicalInstancedStaticMeshComponent>(CookedActor, CookedMeshComponent, InstanceMesh, Name, GeneratedModelHismComponent->GetComponentTransform());
+
+							for (int32 InstanceIndex = 0; InstanceIndex < GeneratedModelHismComponent->GetInstanceCount(); ++InstanceIndex)
+							{
+								FTransform Transform;
+								GeneratedModelHismComponent->GetInstanceTransform(InstanceIndex, Transform);
+								InstancedStaticMeshComponent->AddInstance(Transform);
+							}
+						}

```

Comment:
The if could be moved further inside to reduce code duplication

Reply:
Makes sense thanks!